worker/uniter/pebblewatcher.go

```diff
@@ -0,0 +1,266 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package uniter
+
+import (
+	"context"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/canonical/pebble/client"
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/worker/v3"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/worker/uniter/container"
+)
+
+// PebbleClient describes the subset of Pebble's client.Client that we need
+// for pebbleWatcher.
+type PebbleClient interface {
+	CloseIdleConnections()
+	SysInfo() (*client.SysInfo, error)
+	WaitNotices(ctx context.Context, serverTimeout time.Duration, opts *client.NoticesOptions) ([]*client.Notice, error)
+}
+
+// NewPebbleClientFunc is the function type used to create a PebbleClient.
+type NewPebbleClientFunc func(*client.Config) (PebbleClient, error)
+
+type pebbleWatcher struct {
+	logger            Logger
+	clock             clock.Clock
+	workloadEventChan chan string
+	workloadEvents    container.WorkloadEvents
+	newPebbleClient   NewPebbleClientFunc
+
+	tomb          tomb.Tomb
+	mut           sync.Mutex
+	pebbleBootIDs map[string]string
+}
+
+const (
+	pebblePollInterval = 5 * time.Second
+)
+
+// NewPebbleWatcher starts a worker that watches the Pebble containers for
+// various events: poll for Pebble-ready and wait for Pebble notices.
+func NewPebbleWatcher(
+	logger Logger,
+	clock clock.Clock,
+	containerNames []string,
+	workloadEventChan chan string,
+	workloadEvents container.WorkloadEvents,
+	newPebbleClient NewPebbleClientFunc,
+) worker.Worker {
+	if newPebbleClient == nil {
+		newPebbleClient = func(config *client.Config) (PebbleClient, error) {
+			return client.New(config)
+		}
+	}
+	p := &pebbleWatcher{
+		logger:            logger,
+		clock:             clock,
+		workloadEventChan: workloadEventChan,
+		workloadEvents:    workloadEvents,
+		newPebbleClient:   newPebbleClient,
+		pebbleBootIDs:     make(map[string]string),
+	}
+	for _, v := range containerNames {
+		containerName := v
+		p.tomb.Go(func() error {
+			return p.runPoller(containerName)
+		})
+		p.tomb.Go(func() error {
+			return p.runNoticer(containerName)
+		})
```

Comment:
I think this is not quite right. See doc for `tomb.Go()`
```
// It is safe for the f function to call the Go method again
// to create additional tracked goroutines. Once all tracked
// goroutines return, the Dead channel is closed and the
// Wait method unblocks and returns the death reason.
//
// Calling the Go method after all tracked goroutines return
// causes a runtime panic. For that reason, calling the Go
// method a second time out of a tracked goroutine is unsafe.
```

Reply:
Per discussion, we've reverted this back to two workers, so this is now moot.